bisection.py

```diff
@@ -0,0 +1,373 @@
+    commit_dict: Dict[str, int]
+    def __init__(self, srcpath: str):
+        self.srcpath = srcpath
+        self.commits = []
+        self.commit_dict = dict()
+
+    def prep(self) -> bool:
+        repo_origin_url = gitutils.get_git_origin(self.srcpath)
+        if not repo_origin_url == TORCH_GITREPO:
+            print(f"Unmatched repo origin url: {repo_origin_url} with standard {TORCH_GITREPO}")
+            return False
+        return True
+    
+    # Get all commits between start and end, save them in commits
+    def init_commits(self, start: str, end: str) -> bool:
+        commits = gitutils.get_git_commits(self.srcpath, start, end)
+        if not commits or len(commits) < 2:
+            print(f"Failed to retrieve commits from {start} to {end} in {self.srcpath}.")
+            return False
+        for count, commit in enumerate(commits):
+            ctime = gitutils.get_git_commit_date(self.srcpath, commit)
+            self.commits.append(Commit(sha=commit, ctime=ctime, score=None))
+            self.commit_dict[commit] = count
+        # Setup commit date
+        last_commit_date = self.commits[-1].ctime.split()[0]
+        self.commit_date = datetime.strptime(last_commit_date, "%Y-%m-%d")
+        return True
+    
+    def get_mid_commit(self, left: Commit, right: Commit) -> Optional[Commit]:
+        left_index = self.commit_dict[left.sha]
+        right_index = self.commit_dict[right.sha]
+        if right_index == left_index + 1:
+            return None
+        else:
+            return self.commits[(left_index + right_index) / 2]
+
+    def setup_build_env(self, env):
+        env["USE_CUDA"] = "1"
+        env["BUILD_CAFFE2_OPS"] = "0"
+        env["USE_XNNPACK"] = "0"
+        env["USE_MKLDNN"] = "1"
+        env["USE_MKL"] = "1"
+        env["USE_CUDNN"] = "1"
+        env["CMAKE_PREFIX_PATH"] = env["CONDA_PREFIX"]
+        return env
+
+    # Install dependencies such as torchaudio, torchtext and torchvision
+    def install_deps(self, build_env):
+        # Build torchvision
+        print(f"Building torchvision ...", end="", flush=True)
+        command = "python setup.py install &> /dev/null"
+        subprocess.check_call(command, cwd=TORCHBENCH_DEPS["torchvision"], env=build_env, shell=True)
+        print("done")
+        # Build torchaudio
+        print(f"Building torchaudio ...", end="", flush=True)
+        command = "BUILD_SOX=1 python setup.py install &> /dev/null"
+        subprocess.check_call(command, cwd=TORCHBENCH_DEPS["torchaudio"], env=build_env, shell=True)
+        print("done")
+        # Build torchtext
+        print(f"Building torchtext ...", end="", flush=True)
+        command = "python setup.py clean install &> /dev/null"
+        subprocess.check_call(command, cwd=TORCHBENCH_DEPS["torchtext"], env=build_env, shell=True)
+        print("done")
+ 
+    def build(self, commit: Commit):
+        # checkout pytorch commit
+        gitutils.checkout_git_commit(self.srcpath, commit.sha)
```

Comment:
i'll have to look at checkout_git_commit impl to see, but did you make sure to update submodules recursively when checking out?

Reply:
Yes, I did "git submodule sync && git submodule update --init --recursive" after each checkout.

Reply:
Sometimes there will be untracked files after checkout and it can't be removed, otherwise repo won't compile. Seems no actions needed in this case.